src/main/java/net/neoforged/neoforge/transfer/transaction/TransactionManager.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.transaction;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import java.util.ArrayList;
+import java.util.List;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Manager for handling opening new {@link Transaction Transactions} or querying status of a transaction chain in a given thread.
+ *
+ * @see Transaction
+ */
+final class TransactionManager {
+    private static final ThreadLocal<TransactionManager> MANAGERS = ThreadLocal.withInitial(TransactionManager::new);
+    final Thread thread = Thread.currentThread();
+    final List<Transaction> stack = new ArrayList<>();
+    final List<SnapshotJournal<?>> journalsToCommit = new ArrayList<>();
+    int currentDepth = -1;
+
+    private final Int2ObjectMap<Class<?>> debugMap = new Int2ObjectOpenHashMap<>();
```

Comment:
This would probably make more sense inside each transaction rather than as a map in the transaction manager.